src/librustc_codegen_ssa/mir/rvalue.rs

```diff
@@ -870,36 +869,151 @@ fn cast_float_to_int<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
-        bx.select(cmp, s1, zero)
+    let zero = bx.cx().const_uint(int_ty, 0);
+
+    // The codegen here differes quite a bit depending on whether our builder's
+    // `fptosi` and `fptoui` instructions may trap for out-of-bounds values. If
+    // they don't dtrap then we can start doing everything inline with a
+    // `select` instruction because it's ok to execute `fptosi` and `fptoui`
+    // even if we don't use the results.
+    if !bx.fptosui_may_trap(x, int_ty) {
+        // Step 1 ...
+        let fptosui_result = if signed { bx.fptosi(x, int_ty) } else { bx.fptoui(x, int_ty) };
+        let less_or_nan = bx.fcmp(RealPredicate::RealULT, x, f_min);
+        let greater = bx.fcmp(RealPredicate::RealOGT, x, f_max);
+
+        // Step 2: We use two comparisons and two selects, with %s1 being the
+        // result:
+        //     %less_or_nan = fcmp ult %x, %f_min
+        //     %greater = fcmp olt %x, %f_max
+        //     %s0 = select %less_or_nan, int_ty::MIN, %fptosi_result
+        //     %s1 = select %greater, int_ty::MAX, %s0
+        // Note that %less_or_nan uses an *unordered* comparison. This
+        // comparison is true if the operands are not comparable (i.e., if x is
+        // NaN). The unordered comparison ensures that s1 becomes int_ty::MIN if
+        // x is NaN.
+        //
+        // Performance note: Unordered comparison can be lowered to a "flipped"
+        // comparison and a negation, and the negation can be merged into the
+        // select. Therefore, it not necessarily any more expensive than a
+        // ordered ("normal") comparison. Whether these optimizations will be
+        // performed is ultimately up to the backend, but at least x86 does
+        // perform them.
+        let s0 = bx.select(less_or_nan, int_min, fptosui_result);
+        let s1 = bx.select(greater, int_max, s0);
+
+        // Step 3: NaN replacement.
+        // For unsigned types, the above step already yielded int_ty::MIN == 0 if x is NaN.
+        // Therefore we only need to execute this step for signed integer types.
+        if signed {
+            // LLVM has no isNaN predicate, so we use (x == x) instead
+            let cmp = bx.fcmp(RealPredicate::RealOEQ, x, x);
+            bx.select(cmp, s1, zero)
+        } else {
+            s1
+        }
     } else {
-        s1
+        // In this case we cannot execute `fptosi` or `fptoui` and then later
+        // discard the result. The builder is telling us that these instructions
+        // will trap on out-of-bounds values, so we need to use basic blocks and
+        // control flow to avoid executing the `fptosi` and `fptoui`
+        // instructions.
+        //
+        // The general idea of what we're constructing here is, for f64 -> i32:
+        //
+        //      ;; ... previous basic block
+        //      %load_result = alloca i32, align 4
+        //      %less_or_nan = fcmp ult double %x, 0xC1E0000000000000
```

Comment:
This looks more complicated than it needs to be. I’m aware that this is not very constructive in terms of feedback as it is… but I’m experimenting with some alternative ways to write this down as LLVM IR and hopefully will come back to update this before monday.

Reply:
Here’s something I have so far (caveat untested, unchecked and not validated):

```llvm
  ; check that the number is within the expected bounds and just convert if the number is in range
  %inbound_lower = fcmp oge double %0, 0xC1E0000000000000
  %inbound_upper = fcmp ole double %0, 0x41DFFFFFFFC00000
  %inbounds = and i1 %inbound_lower, %inbound_upper
  br i1 %inbounds, label %convert, label %specialcase

specialcase:
  %isnan = fcmp ord double %0, 0.000000e+00 ; can probably also be `!inbound_lower && !inbound_upper`?
  ; select a max/min value for the case where float is not nan
  %result_notnan = select i1 %inbound_lower, i32 2147483647, i32 -2147483648
  ; nan or min/max
  %result_nan = select i1 %isnan, i32 0, i32 %result_notnan
  ret i32 %result_nan

convert:
  %result = fptosi double %0 to i32
  ret i32 %result
```

Reply:
This is what I ended up in the end

```llvm
    ;; block so far... %0 is the argument
    %inbound_lower = fcmp oge double %0, 0xC1E0000000000000
    %inbound_upper = fcmp ole double %0, 0x41DFFFFFFFC00000
    ;; match (inbound_lower, inbound_upper) { 
    ;;     (true, true) => %0 can be converted without trapping
    ;;     (false, false) => %0 is a NaN
    ;;     (true, false) => %0 is too large
    ;;     (false, true) => %0 is too small
    ;; }
    ;;
    ;; The (true, true) check, go to %convert if so.
    %inbounds = and i1 %inbound_lower, %inbound_upper
    br i1 %inbounds, label %convert, label %specialcase

specialcase:
    ;; Handle the cases where the number is NaN, too large or too small
    
    ;; Either (true, false) or (false, true)
    %is_not_nan = or i1 %inbound_lower, %inbound_upper
    ;; Figure out which saturated value we are interested in if not `NaN`
    %saturated = select i1 %inbound_lower, i32 2147483647, i32 -2147483648
    ;; Figure out between saturated and NaN representations
    %result_nan = select i1 %is_not_nan, i32 %saturated, i32 0
    br label %done

convert:
    %result = call i32 @llvm.wasm.trunc.signed.i32.f64(double %0) ; fptosi double %0 to i32
    br label %done

done:
    %r = phi i32 [ %result, %convert ], [ %result_nan, %specialcase ]
    ;; continues here
```

Here’s a godbolt comparison of the two approaches: https://rust.godbolt.org/z/cnEGWM. The new one looks somewhat better to my untrained eye, and its also something LLVM is not really able to optimize much further, so at the very least there’s a very minor win in terms of compile time.

Reply:
Oh nice! I'm happy to update this to use that codegen. Do you know if we can natively generate phi nodes though? I wanted to do that instead of using an `alloca` but the builder trait didn't seem to have a way to manually construct phi nodes.